### DISCUSSION TITLE
Signing bundletool build-apks with p12 vs keystore.ks?

### DISCUSSION THREAD

**Discussion by FliplineStudios** [airsdk/Adobe-Runtime-Support#1601] [Q&A]
I've been going through the process of packaging an AAB for the first time, going through the tutorial on airsdk.dev to walk through the process.  After upgrading Java I've been able to build an aab, and I'm working through testing the app locally with bundletool.

Before I realized that airsdk.dev mentioned creating a keystore.ks for `bundletool build-apks`, I ended up just using my regular p12 certificate for the --ks parameter, and the regular alias that keytool gave me for that p12 for --ks-key-alias.

As best I can tell, it seemed to have worked successfully without requiring a keystore.ks -- I'm able to install it with `bundletool install-apks` without an issue, the app runs on the device, and checking out the RSA files within each APK in the APKS package, keytool is showing all of our info from our p12 certificate.

I'm just curious if this is a good idea using the p12 in build-apks or not, and if there's any benefit to using a keystore ks instead, or any downsides to not using it.  I'm not particularly well versed in signing or keystores, having never had to package anything besides letting adt do all the work in the past, so any insight would be appreciated!

**Comment by marchbold** [comment]
I've come to the same conclusion, that as long as the p12 contains the same format as the required keystore that it works fine. We have actually just used a p12 generated from adt for all of the process. They are just slightly different containers for the same certificate so not surprising that it works :)

The only part I haven't confirmed yet is the upload of the "upload certificate" to the play console. Not sure if that needs to be a `keystore` file as yet. Did you try to upload a `p12` file ?

**Comment by FliplineStudios** [comment]
@marchbold Great thanks!

I haven't tried uploading to Google Play yet, that's actually next on my list now that I've tested the aab/apks locally.  If I'm understanding the tutorial correctly, we'll have to use the PEPK tool to create the ZIP from our existing p12 for Play App Signing, but the Upload Certificate is optional and we can just skip that?  Or is it optional only for older existing apps, but brand new apps are required to have a separate upload key?

**Reply by marchbold** [reply]
If you have a 1024bit certificate then definitely create an upload certificate (use a 2048 bit one so you can get around using the old java version in a crazy complex setup). 

Otherwise I haven't found a good reason other than if you have external developers or a large company where you don't want to potentially expose your app certificate. 

Once you have uploaded your app certificate you can always change the upload certificate later. I suppose it's just that additional layer of security?
